### PR TITLE
Performance monitoring counters 

### PR DESCRIPTION
**Pull request by jmcadden**

### CODE REVIEW

File: baremetal/src/Perf.cc [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+#include <ebbrt/CpuAsm.h>
+#include <ebbrt/Cpuid.h>
+#include <ebbrt/Debug.h>
+#include <ebbrt/Perf.h>
+
+namespace {
+thread_local uint8_t pmcs{0};  // stores allocated gp counters
+};
+
+thread_local uint64_t ebbrt::perf::perf_global_ctrl{0};  // stored activated counters
```
Comment (dschatzberg) [inline baremetal/src/Perf.cc]: no need to put either of these two initializers. Zero is the default value for these already.

File: baremetal/src/Perf.cc [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  default:
+    perfevtsel.usermode = 0;
+    perfevtsel.osmode = 0;
+    perfevtsel.en = 0;
+    ebbrt::wrmsr(perfevtsel.val, IA32_PERFEVTSEL_MSR(pmc_num_));
+    pmcs &= ~(1ull << pmc_num_);
+    break;
+  };
+
+  return;
```
Comment (dschatzberg) [inline baremetal/src/Perf.cc]: this return is superfluous

File: baremetal/src/Perf.cc [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        kprintf("Perf counter #%d initialized to evt=%u\n", pmc_num_, static_cast<uint8_t>(evt_));
+        return;
+      }
+    }
+    kprintf("Warning: no available hardware counters.\n");
+    evt_ = PerfEvent::nil_event;
+  }
+
+
+  return;
```
Comment (dschatzberg) [inline baremetal/src/Perf.cc]: this return is superfluous

File: baremetal/src/Perf.cc [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    counter_offset_ = ebbrt::rdmsr(IA32_FXD_PMC(1));
+    break;
+  case PerfEvent::fixed_reference_cycles:
+    counter_offset_ = ebbrt::rdmsr(IA32_FXD_PMC(2));
+    break;
+  default:
+    counter_offset_ = ebbrt::rdmsr(IA32_PMC(pmc_num_));
+    break;
+  };
+  return;
```
Comment (dschatzberg) [inline baremetal/src/Perf.cc]: this return is superfluous

File: baremetal/src/include/ebbrt/CpuAsm.h [hunk 5]
```diff
@@ -5,12 +5,46 @@
...
+  uint64_t tsc;
+  asm volatile("rdtscp;"
+               "shl $32,%%rdx;"
+               "or %%rdx,%%rax"
+               : "=a"(tsc)
+               :
+               : "%rcx", "%rdx");
+  return tsc;
+}
+
```
Comment (dschatzberg) [inline baremetal/src/include/ebbrt/CpuAsm.h]: These already exist in Rdtsc.h

File: baremetal/src/include/ebbrt/Perf.h [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+#define PERFEVTSEL_UMASK_BRANCH_INSTRUCTIONS 0x00
+// branch misses
+#define PERFEVTSEL_EVT_BRANCH_MISSES 0xC5
+#define PERFEVTSEL_UMASK_BRANCH_MISSES 0x00
+// (Intel Xeon) TLB miss all levels
+#define PERFEVTSEL_EVT_TLB_STORE_MISSES 0x49
+#define PERFEVTSEL_UMASK_TLB_STORE_MISSES 0x0E
+
+#define PERFEVTSEL_EVT_TLB_LOAD_MISSES 0x08
+#define PERFEVTSEL_UMASK_TLB_LOAD_MISSES 0x01
```
Comment (dschatzberg) [inline baremetal/src/include/ebbrt/Perf.h]: I'd rather see all these typed correctly like so:
`const constexpr uint8_t kUmaskTlbLoadMisses = 0x01;`

Also the function macros can be made into constexpr functions.

File: baremetal/src/include/ebbrt/Perf.h [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+      uint64_t pin : 1;
+      uint64_t enint : 1;
+      uint64_t reserved2 : 1;
+      uint64_t en : 1;
+      uint64_t inv : 1;
+      uint64_t cntmask : 8;
+      uint64_t reserved0 : 32;
+    } __attribute__((packed));
+  };
+} ia32_perfevtsel_t;
```
Comment (dschatzberg) [inline baremetal/src/include/ebbrt/Perf.h]: for our own types, use CamelCase, also can you put in a comment referencing where you got this from?

File: baremetal/src/include/ebbrt/Perf.h [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    return;
+  }
+  void Stop() __attribute__((always_inline)) {
+    if (evt_ != PerfEvent::nil_event) {
+      perf_global_ctrl &= ~(1ull << pmc_num_);
+      ebbrt::wrmsr(perf_global_ctrl, IA32_PERF_GLOBAL_CTRL_MSR);
+    }
+    return;
+  }
+  int8_t AvailablePMCs();
```
Comment (dschatzberg) [inline baremetal/src/include/ebbrt/Perf.h]: this always returns zero? What is the point? Shouldn't it be static?

Comment (jmcadden) [inline baremetal/src/include/ebbrt/Perf.h]: For the time being I've removed the unimplemented method

**Comment by jmcadden** [comment]
@BetsyRT r+

**Comment by jmcadden** [comment]
@BetsyRT try

**Comment by dschatzberg** [comment]
@BetsyRT try

**Comment by BetsyRT** [comment]
:hourglass: Trying commit 81722f0 with merge 69058d0...

**Comment by BetsyRT** [comment]
:boom: Test timed out

**Comment by jmcadden** [comment]
@BetsyRT retry

**Comment by dschatzberg** [comment]
@BetsyRT retry

**Comment by jmcadden** [comment]
@BetsyRT try

**Comment by dschatzberg** [comment]
@BetsyRT clean

File: baremetal/src/include/ebbrt/CpuAsm.h [hunk 9]
```diff
@@ -5,12 +5,24 @@
...
+inline void wrmsr(uint64_t val, uint32_t msr) {
+  asm volatile("wrmsr;" : : "a"(val & 0xFFFFFFFF), "d"(val >> 32), "c"(msr));
+}
+
+inline uint64_t rdmsr(uint32_t msr) {
+  uint64_t val;
+  asm volatile("rdmsr;" : "=A"(val) : "c"(msr));
+  return val;
+}
```
Comment (dschatzberg) [inline baremetal/src/include/ebbrt/CpuAsm.h]: we have Msr.h already with Read and Write

**Comment by dschatzberg** [comment]
@BetsyRT force

**Comment by BetsyRT** [comment]
:bomb: Buildbot returned an error: `Authorization failed`

**Comment by dschatzberg** [comment]
@BetsyRT clean

**Comment by dschatzberg** [comment]
@BetsyRT retry

**Pull request closed without merging**